### PR TITLE
[Merged by Bors] - Handle early blocks

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Problem this PR addresses

There's an issue where Lighthouse is banning a lot of peers due to the following sequence of events:

1. Gossip block 0xabc arrives ~200ms early
    - It is propagated across the network, with respect to [`MAXIMUM_GOSSIP_CLOCK_DISPARITY`](https://github.com/ethereum/eth2.0-specs/blob/v1.0.0/specs/phase0/p2p-interface.md#why-is-there-maximum_gossip_clock_disparity-when-validating-slot-ranges-of-messages-in-gossip-subnets).
    - However, it is not imported to our database since the block is early.
2. Attestations for 0xabc arrive, but the block was not imported.
    - The peer that sent the attestation is down-voted.
        - Each unknown-block attestation causes a score loss of 1, the peer is banned at -100.
        - When the peer is on an attestation subnet there can be hundreds of attestations, so the peer is banned quickly (before the missed block can be obtained via rpc).

## Potential solutions

I can think of three solutions to this:

1. Wait for attestation-queuing (#635) to arrive and solve this.
    - Easy
    - Not immediate fix.
    - Whilst this would work, I don't think it's a perfect solution for this particular issue, rather (3) is better.
1. Allow importing blocks with a tolerance of `MAXIMUM_GOSSIP_CLOCK_DISPARITY`.
    - Easy
    - ~~I have implemented this, for now.~~
1. If a block is verified for gossip propagation (i.e., signature verified) and it's within `MAXIMUM_GOSSIP_CLOCK_DISPARITY`, then queue it to be processed at the start of the appropriate slot.
    - More difficult
    - Feels like the best solution, I will try to implement this.
    
    
**This PR takes approach (3).**

## Changes included

- Implement the `block_delay_queue`, based upon a [`DelayQueue`](https://docs.rs/tokio-util/0.6.3/tokio_util/time/delay_queue/struct.DelayQueue.html) which can store blocks until it's time to import them.
- Add a new `DelayedImportBlock` variant to the `beacon_processor::WorkEvent` enum to handle this new event.
- In the `BeaconProcessor`, refactor a `tokio::select!` to a struct with an explicit `Stream` implementation. I experienced some issues with `tokio::select!` in the block delay queue and I also found it hard to debug. I think this explicit implementation is nicer and functionally equivalent (apart from the fact that `tokio::select!` randomly chooses futures to poll, whereas now we're deterministic).
- Add a testing framework to the `beacon_processor` module that tests this new block delay logic. I also tested a handful of other operations in the beacon processor (attns, slashings, exits) since it was super easy to copy-pasta the code from the `http_api` tester.
    - To implement these tests I added the concept of an optional `work_journal_tx` to the `BeaconProcessor` which will spit out a log of events. I used this in the tests to ensure that things were happening as I expect.
    - The tests are a little racey, but it's hard to avoid that when testing timing-based code. If we see CI failures I can revise. I haven't observed *any* failures due to races on my machine or on CI yet.
    - To assist with testing I allowed for directly setting the time on the `ManualSlotClock`.
- I gave the `beacon_processor::Worker` a `Toolbox` for two reasons; (a) it avoids changing tons of function sigs when you want to pass a new object to the worker and (b) it seemed cute.

### CODE REVIEW

**Comment by paulhauner** [comment]
> Each unknown-block attestation causes a score loss of 1, the peer is banned at -100.
>
>  When the peer is on an attestation subnet there can be hundreds of attestations, so the peer is banned quickly (before the missed block can be obtained via rpc).

This suggests that we're being too harsh on unknown-block attestations, but perhaps I'm missing something. Do you have thoughts on this @divagant-martian? :)

**Comment by AgeManning** [comment]
This is supposed to be handled by caching attestations. 

Each attestation with an unknown block should get cached and a request via RPC to find the block should take place. We should only penalize after the peer fails to send the block via the RPC. #635 - @divagant-martian was working on this

**Comment by paulhauner** [comment]
> This is supposed to be handled by caching attestations.

I agree it would help us here (and many other places), but for this specific scenario I think queuing blocks is a more efficient solution in the long term. No need for an RPC call to re-get the block and no need to use memory to cache attestations.

**Comment by paulhauner** [comment]
~~FYI I haven't seen this pass CI yet, but it works locally. ~~

Not sure why ~~strikethrough~~ isn't working above... But consider that statement retracted.

File: beacon_node/network/src/beacon_processor/block_delay_queue.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
+//! Provides a mechanism which queues blocks for later processing when they arrive too early.
+//!
+//! When the `beacon_processor::Worker` imports a block that is acceptably early (i.e., within the
+//! gossip propagation tolerance) it will send it to this queue where it will be placed in a
+//! `DelayQueue` until the slot arrives. Once the block has been determined to be ready, it will be
+//! sent back out a channel to be processed by the `BeaconProcessor` again.
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/block_delay_queue.rs]:
```suggestion
//! sent back out on a channel to be processed by the `BeaconProcessor` again.
```

File: beacon_node/network/src/beacon_processor/block_delay_queue.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        let mut queued_block_roots = HashSet::new();
+
+        let mut inbound_events = InboundEvents {
+            early_blocks_rx,
+            delay_queue: DelayQueue::new(),
+        };
+
+        loop {
+            match inbound_events.next().await {
+                // Some block has been indicated as "early" and should be processed with the
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/block_delay_queue.rs]:
```suggestion
                // Some block has been indicated as "early" and should be processed when the
```

File: beacon_node/network/src/beacon_processor/block_delay_queue.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                    let block_root = early_block.block.block_root;
+
+                    // Don't add the same block to the queue twice. This prevents DoS attacks.
+                    if queued_block_roots.contains(&block_root) {
+                        continue;
+                    }
+
+                    if let Some(duration_till_slot) = slot_clock.duration_to_slot(block_slot) {
+                        // Check to ensure this won't over-fill the queue.
+                        if queued_block_roots.len() > MAXIMUM_QUEUED_BLOCKS {
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/block_delay_queue.rs]: Should this be `>=`?

File: beacon_node/network/src/beacon_processor/tests.rs [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+                ),
+            }
+        })
+    }
+
+    /// Assert that the `BeaconProcessor` event journal is as `expected`.
+    ///
+    /// ## Note
+    ///
+    /// We wont attempt to listen for an more than `expected.len()` events. As such, it makes sense
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/tests.rs]:
```suggestion
    /// We won't attempt to listen for any more than `expected.len()` events. As such, it makes sense
```

File: beacon_node/network/src/beacon_processor/tests.rs [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+    // and check the head in the time between the block arrived early and when its due for
+    // processing.
+    //
+    // If this causes issues we might be able to make the block delay queue add a longer delay for
+    // processing, instead of just MAXIMUM_GOSSIP_CLOCK_DISPARITY. Speak to @paulhauner if this test
+    // starts failing.
+    rig.chain.slot_clock.set_slot(rig.next_block.slot().into());
+    assert!(
+        rig.chain.head().unwrap().beacon_block_root != rig.next_block.canonical_root(),
+        "block not yet be imported"
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/tests.rs]:
```suggestion
        "block not yet imported"
```

File: beacon_node/network/src/beacon_processor/tests.rs [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+    rig.assert_event_journal(&[DELAYED_IMPORT_BLOCK, WORKER_FREED, NOTHING_TO_DO]);
+
+    assert_eq!(
+        rig.chain.head().unwrap().beacon_block_root,
+        rig.next_block.canonical_root(),
+        "block should be imported and become head"
+    );
+}
+
+/// Blocks that are *too* early shouldn't got into the delay queue.
```
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/tests.rs]:
```suggestion
/// Blocks that are *too* early shouldn't get into the delay queue.
```

**Review by michaelsproul — COMMENTED** [review]
Looks good!

Great work on the tests particularly, I know how tricksy they were, but they're really valuable :heart: 

Just a few typo corrections and then we're ready to merge I think!

File: beacon_node/network/src/beacon_processor/block_delay_queue.rs (hunk 3 above)
Comment (paulhauner) [reply beacon_node/network/src/beacon_processor/block_delay_queue.rs]: Off by one! Nice catch :pray:

**Comment by paulhauner** [comment]
All comments addressed! Thank you!

**Review by michaelsproul — APPROVED** [review]
:ok_hand:

**Pull request closed without merging**